quantum/process_keycode/process_leaders.c

```diff
@@ -0,0 +1,177 @@
+/* Copyright 2016 Alex Olkhovskiy
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef LEADERS_ENABLE
+
+#include "process_leaders.h"
+
+#include "print.h"
+
+uint16_t momentary_sentinels;
+bool oneshot_sentinel;
+uint16_t keyseq_codes[LEADERS_SEQ_MAX];
+
+uint16_t** keyseq_definitions;
+
+uint8_t keyseq_index;
+uint8_t ref_layer;
+
+__attribute__ ((weak))
+void leaders_init_user(void) {
+}
+
+__attribute__ ((weak))
+void keyseq_set_definitions(uint16_t* k[]) {
+  keyseq_definitions = k;
+}
+
+__attribute__ ((weak))
+bool keyseq_press_user(uint16_t keycode, keyrecord_t *record) { return true;}
+
+__attribute__ ((weak))
+void keyseq_release_user(uint16_t keycode) {
+}
+
+void leaders_init() {
+  init_press_state();
+  momentary_sentinels = 0x0000;
+  oneshot_sentinel = true;
+  keyseq_index = 0;
+  leaders_init_user();
+}
+
+void keyseq_push(uint16_t keycode) {
+  momentary_sentinels |= (1U << keyseq_index );
+  keyseq_codes[keyseq_index++] = keycode;
+}
+
+void keyseq_reset_oneshot() {
+  if (momentary_sentinels == 0) {
+    oneshot_sentinel = true;
+    keyseq_index = 0;
+  } else {
+    oneshot_sentinel = false;
+    keyseq_index = 1 + biton16(momentary_sentinels);
+  }
+  return ;
+}
+
+void keyseq_set_oneshot() {
+  oneshot_sentinel = true;
+}
+
+void keyseq_reset_momentary(uint8_t pos) {
+  momentary_sentinels &= ~(1U << (pos - 1));
+  if (oneshot_sentinel) {
+    return;
+  }
+  if (momentary_sentinels == 0) {
+    keyseq_index = 0;
+    oneshot_sentinel = true;
+  } else {
+    keyseq_index = 1 + biton16(momentary_sentinels);
+  }
+  return ;
+}
+
+bool process_leaders(uint16_t keycode, keyrecord_t *record) {
+  if (keycode == KC_LCTL ||
+      keycode == KC_RCTL ||
+      keycode == KC_LGUI ||
+      keycode == KC_RGUI ||
+      keycode == KC_LALT ||
+      keycode == KC_RALT ) {
+    return true;
```

Comment:
Do I understand correctly that this means mods can't be part of a leader sequence? Can you explain why that is?

Reply:
There is nothing that stops them to be part of a sequence.  I use sequences to switch between layers and I found it very useful to combine them with modifiers. 
But it is a matter of personal preference and it can be made configurable.